File: src/lib/ai-fixes.ts
import Anthropic from '@anthropic-ai/sdk'
import type { AuditCheck } from './types'

interface Fix {
  id: string
  fix_code?: string
  fix_explanation?: string
  fix_location?: string
}

export async function generateFixes(
  url: string,
  html: string,
  failedChecks: AuditCheck[],
): Promise<Fix[]> {
  const apiKey = process.env.ANTHROPIC_API_KEY
  if (!apiKey) return []

  const client = new Anthropic({ apiKey })

  // Keep only <head> to stay within token limits
  const headEnd = html.indexOf('</head>')
  const truncated = headEnd > 0 ? html.slice(0, headEnd + 7) : html.slice(0, 3000)

  const checkList = failedChecks
    .map(c => `- ${c.id}: ${c.name} — ${c.description}`)
    .join('\n')

  const response = await client.messages.create({
    model: 'claude-sonnet-4-5-20250514',
    max_tokens: 2000,
    messages: [
      {
        role: 'user',
        content: `You are a web SEO expert. A website at ${url} has these issues:

${checkList}

Current HTML <head>:
\`\`\`html
${truncated}
\`\`\`

For each failed check, return a JSON array of objects with:
- "id": check ID (exact match)
- "fix_code": exact HTML/XML to copy-paste
- "fix_explanation": 1-2 sentence plain-English explanation
- "fix_location": where to put it

For descriptions: write based on actual page content, 120-160 chars, include keywords.
For JSON-LD: generate Organization schema from page content.
For sitemaps: generate complete sitemap.xml with today's date.

Return ONLY a JSON array. No markdown fences.`,
      },
    ],
  })

  const text = (response.content[0] as { type: 'text'; text: string }).text.trim()

  // Strip markdown fences if present
  let json = text
  if (json.startsWith('```')) {
    json = json.split('\n').slice(1).join('\n')
    if (json.endsWith('```')) json = json.slice(0, -3).trim()
  }

  try {
    const fixes = JSON.parse(json)
    if (Array.isArray(fixes)) {
      const validIds = new Set(failedChecks.map(c => c.id))
      return fixes.filter((f): f is Fix => typeof f === 'object' && validIds.has(f.id))
    }
  } catch {
    // AI response wasn't valid JSON — acceptable
  }

  return []
}

File: backend/app/services/ai_fixes.py
"""
AI Fix Generator — Uses Claude API to generate copy-paste fix code
for failed audit checks based on the actual page content.
"""

import os
import json

from anthropic import AsyncAnthropic


async def generate_fixes(url: str, html: str, failed_checks: list[dict]) -> list[dict]:
    """Generate fix code for failed checks using Claude AI."""
    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        return []

    client = AsyncAnthropic(api_key=api_key)

    # Truncate HTML to avoid token limits (keep head + first 2000 chars of body)
    head_end = html.find("</head>")
    if head_end > 0:
        truncated_html = html[: head_end + 7] + "\n<!-- body truncated -->"
    else:
        truncated_html = html[:3000]

    check_ids = [c["id"] for c in failed_checks]
    check_descriptions = "\n".join(
        f"- {c['id']}: {c['name']} — {c['description']}"
        for c in failed_checks
    )

    prompt = f"""You are a web SEO expert. A website at {url} has the following issues from an automated audit.

For each failed check below, generate the exact fix code the user should add to their website.

## Failed checks:
{check_descriptions}

## Current HTML <head>:
```html
{truncated_html}
```

## Instructions:
For each failed check, return a JSON object with:
- "id": the check ID (must match exactly)
- "fix_code": the exact HTML/XML/text code to add or replace (ready to copy-paste)
- "fix_explanation": a 1-2 sentence plain-English explanation of what this fix does and why it matters
- "fix_location": where to put the code (e.g., "Inside <head> tag", "Root folder as sitemap.xml")

For meta descriptions and OG descriptions: write them based on the actual page content. Make them compelling, 120-160 characters, and include relevant keywords.

For JSON-LD: generate an Organization schema with the business name and URL extracted from the page.

For sitemaps: generate a complete sitemap.xml with the URL and today's date as lastmod.

Return ONLY a JSON array of fix objects. No markdown, no explanation outside the JSON."""

    response = await client.messages.create(
        model="claude-sonnet-4-5-20250514",
        max_tokens=2000,
        messages=[{"role": "user", "content": prompt}],
    )

    # Parse the response
    text = response.content[0].text.strip()

    # Strip markdown code fences if present
    if text.startswith("```"):
        text = text.split("\n", 1)[1]
        if text.endswith("```"):
            text = text[:-3].strip()

    try:
        fixes = json.loads(text)
        if isinstance(fixes, list):
            return [f for f in fixes if isinstance(f, dict) and f.get("id") in check_ids]
    except json.JSONDecodeError:
        pass

    return []
